### ISSUE TITLE
Missed Detection in Ultralytics vs GitHub/HF weights

### ISSUE THREAD

**Issue by reachpranjal** [ultralytics/ultralytics#18331]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

I find some differences in inference results between Ultralytics release and GitHub weights. I'm referring only to the **_"building"_** text prompt which the ultralytics version completely misses for all the v2 (s, m, l, x) weights.

### Ultralytics Weights + API
![tmp9jfmgwrl](https://github.com/user-attachments/assets/1b085cbd-a932-44ea-97dd-db306b72efb4)

### GitHub Weights or Huggingface Weights (Source Code)
![image](https://github.com/user-attachments/assets/d60742db-09db-4bf8-8798-e81bc3c4942b)

### Environment

Ultralytics 8.3.51 🚀 Python-3.11.11 torch-2.4.0+cu121 CUDA:0 (NVIDIA GeForce RTX 4090 Laptop GPU, 16072MiB)
Setup complete ✅ (20 CPUs, 31.0 GB RAM, 1133.6/1482.9 GB disk)

```
OS                  Linux-6.8.0-76060800daily20240311-generic-x86_64-with-glibc2.35
Environment         Linux
Python              3.11.11
Install             pip
RAM                 30.97 GB
Disk                1133.6/1482.9 GB
CPU                 13th Gen Intel Core(TM) i9-13900H
CPU count           20
GPU                 NVIDIA GeForce RTX 4090 Laptop GPU, 16072MiB
GPU count           1
CUDA                12.1

numpy               ✅ 1.26.4>=1.23.0
numpy               ✅ 1.26.4<2.0.0; sys_platform == "darwin"
matplotlib          ✅ 3.9.4>=3.3.0
opencv-python       ✅ 4.9.0>=4.6.0
pillow              ✅ 10.3.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.14.1>=1.4.1
torch               ✅ 2.4.0+cu121>=1.8.0
torch               ✅ 2.4.0+cu121!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.19.0+cu121>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 6.1.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.2>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.13>=2.0.0
```

### Minimal Reproducible Example

```
from ultralytics import YOLO

# Initialize a YOLO-World model
model = YOLO("yolov8l-worldv2.pt")

# Define custom classes
model.set_classes(["building", "traffic sign"])

# Execute prediction for specified categories on an image
results = model.predict("image.jpg")

# Show results
results[0].show()
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @reachpranjal, thank you for your interest in Ultralytics 🚀! We appreciate your detailed report, including the comparison images, environment details, and minimal reproducible example. 

We recommend first ensuring your setup is fully up-to-date, as this often resolves discrepancies:

## Upgrade

Upgrade to the latest `ultralytics` package with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

Once upgraded, retry your example to see if the issue persists.

---

### 🐛 Bug Report

If this is a potential bug, thank you for including a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). This is very helpful for debugging! To further assist, could you clarify the following:
1. Are there consistent cases where the results differ significantly between the two weights?
2. Have you verified that the input data (`image.jpg`) and preprocessing steps are identical between the two tests?

Additionally, slight variations in model performance can stem from differences in weight initialization, training processes, or postprocessing configurations. For a deeper dive, we recommend ensuring identical parameter settings for both versions during inference.

---

### Resources

For new Ultralytics users, we recommend exploring our [Docs](https://docs.ultralytics.com), particularly these usage examples:
- [Python](https://docs.ultralytics.com/usage/python/)
- [CLI](https://docs.ultralytics.com/usage/cli/)

Also, for custom training or usage questions, check out our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) guide.

You can connect with the Ultralytics community to exchange ideas and share experiences:
- Real-time chat on [Discord](https://discord.com/invite/ultralytics) 🎧
- Discussions on [Discourse](https://community.ultralytics.com)
- Threads on the [Subreddit](https://reddit.com/r/Ultralytics)

---

### Environments

Run YOLO in any of the following verified environments with all dependencies pre-installed (including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/)):
- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM ([GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/))
- **AWS** Deep Learning AMI ([AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/))
- **Docker Images** ([Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/)) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

---

### Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI tests](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) are currently passing. These tests verify functionality across YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu.

---

This is an automated response to guide you toward resolving your issue as quickly as possible. An Ultralytics engineer will review your case and provide additional assistance soon 😊

**Comment by glenn-jocher** [comment]
@reachpranjal thank you for reporting this. The differences in detection could stem from variations in weights architecture or training data between Ultralytics and third-party repositories like GitHub or Hugging Face. For better performance, ensure you're using the latest Ultralytics weights and API. If the issue persists, consider adjusting the confidence threshold, image size, or training a custom model fine-tuned to your specific prompts. For further guidance, refer to our [YOLO-World documentation](https://docs.ultralytics.com/models/yolo-world/).
